docs/man/man3/Texture.3
.TH "Texture" 3 "Tue Nov 24 2015" "Version 0.0.0.1" "Fusion3D" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Texture \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <texture\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBTexture\fP (const std::string &fileName, GLenum textureTarget=GL_TEXTURE_2D, GLfloat filter=GL_LINEAR_MIPMAP_LINEAR, GLenum internalFormat=GL_RGBA, GLenum format=GL_RGBA, bool clamp=false, GLenum attachment=GL_NONE)"
.br
.ti -1c
.RI "\fBTexture\fP (int width=0, int height=0, unsigned char *data=0, GLenum textureTarget=GL_TEXTURE_2D, GLfloat filter=GL_LINEAR_MIPMAP_LINEAR, GLenum internalFormat=GL_RGBA, GLenum format=GL_RGBA, bool clamp=false, GLenum attachment=GL_NONE)"
.br
.ti -1c
.RI "\fBTexture\fP (const \fBTexture\fP &texture)"
.br
.ti -1c
.RI "void \fBoperator=\fP (\fBTexture\fP texture)"
.br
.ti -1c
.RI "virtual \fB~Texture\fP ()"
.br
.ti -1c
.RI "void \fBBind\fP (unsigned int unit=0) const "
.br
.ti -1c
.RI "void \fBBindAsRenderTarget\fP () const "
.br
.ti -1c
.RI "int \fBGetWidth\fP () const "
.br
.ti -1c
.RI "int \fBGetHeight\fP () const "
.br
.ti -1c
.RI "bool \fBoperator==\fP (const \fBTexture\fP &texture) const "
.br
.ti -1c
.RI "bool \fBoperator!=\fP (const \fBTexture\fP &texture) const "
.br
.in -1c
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Texture::Texture (const std::string & fileName, GLenum textureTarget = \fCGL_TEXTURE_2D\fP, GLfloat filter = \fCGL_LINEAR_MIPMAP_LINEAR\fP, GLenum internalFormat = \fCGL_RGBA\fP, GLenum format = \fCGL_RGBA\fP, bool clamp = \fCfalse\fP, GLenum attachment = \fCGL_NONE\fP)"

.SS "Texture::Texture (int width = \fC0\fP, int height = \fC0\fP, unsigned char * data = \fC0\fP, GLenum textureTarget = \fCGL_TEXTURE_2D\fP, GLfloat filter = \fCGL_LINEAR_MIPMAP_LINEAR\fP, GLenum internalFormat = \fCGL_RGBA\fP, GLenum format = \fCGL_RGBA\fP, bool clamp = \fCfalse\fP, GLenum attachment = \fCGL_NONE\fP)"

.SS "Texture::Texture (const \fBTexture\fP & texture)"

.SS "Texture::~Texture ()\fC [virtual]\fP"

.SH "Member Function Documentation"
.PP 
.SS "void Texture::Bind (unsigned int unit = \fC0\fP) const"

.SS "void Texture::BindAsRenderTarget () const"

.SS "int Texture::GetHeight () const\fC [inline]\fP"

.SS "int Texture::GetWidth () const\fC [inline]\fP"

.SS "bool Texture::operator!= (const \fBTexture\fP & texture) const\fC [inline]\fP"

.SS "void Texture::operator= (\fBTexture\fP texture)"

.SS "bool Texture::operator== (const \fBTexture\fP & texture) const\fC [inline]\fP"


.SH "Author"
.PP 
Generated automatically by Doxygen for Fusion3D from the source code\&.


src/rendering/texture.cpp
/*
 * Copyright (C) 2015 Dominik Madarasz
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "texture.h"

#include "../core/util.h"
#include "../core/math3d.h"
#include "../core/profiling.h"

#include "../staticLibs/stb_image.h"

#include <iostream>
#include <cassert>
#include <cstring>

std::map<std::string, TextureData*> Texture::s_resourceMap;

TextureData::TextureData(GLenum textureTarget, int width, int height, int numTextures, unsigned char** data, GLfloat* filters, GLenum* internalFormat, GLenum* format, bool clamp, GLenum* attachments)
{
	m_textureID = new GLuint[numTextures];
	m_textureTarget = textureTarget;
	m_numTextures = numTextures;
	
	#if PROFILING_SET_2x2_TEXTURE == 0
		m_width = width;
		m_height = height;
	#else
		m_width = 2;
		m_height = 2;
	#endif
	m_frameBuffer = 0;
	m_renderBuffer = 0;
	
	InitTextures(data, filters, internalFormat, format, clamp);
	InitRenderTargets(attachments);
}

TextureData::~TextureData()
{
	if(*m_textureID) glDeleteTextures(m_numTextures, m_textureID);
	if(m_frameBuffer) glDeleteFramebuffers(1, &m_frameBuffer);
	if(m_renderBuffer) glDeleteRenderbuffers(1, &m_renderBuffer);
	if(m_textureID) delete[] m_textureID;
}

void TextureData::InitTextures(unsigned char** data, GLfloat* filters, GLenum* internalFormat, GLenum* format, bool clamp)
{
	glGenTextures(m_numTextures, m_textureID);
	for(int i = 0; i < m_numTextures; i++)
	{
		glBindTexture(m_textureTarget, m_textureID[i]);
			
		glTexParameterf(m_textureTarget, GL_TEXTURE_MIN_FILTER, filters[i]);
		glTexParameterf(m_textureTarget, GL_TEXTURE_MAG_FILTER, filters[i]);
		
		if(clamp)
		{
			glTexParameterf(m_textureTarget, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
			glTexParameterf(m_textureTarget, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
		}
		
		glTexImage2D(m_textureTarget, 0, internalFormat[i], m_width, m_height, 0, format[i], GL_UNSIGNED_BYTE, data[i]);
		
		if(filters[i] == GL_NEAREST_MIPMAP_NEAREST ||
			filters[i] == GL_NEAREST_MIPMAP_LINEAR ||
			filters[i] == GL_LINEAR_MIPMAP_NEAREST ||
			filters[i] == GL_LINEAR_MIPMAP_LINEAR)
		{
			glGenerateMipmap(m_textureTarget);
			GLfloat maxAnisotropy;
			glGetFloatv(GL_MAX_TEXTURE_MAX_ANISOTROPY_EXT, &maxAnisotropy);
			glTexParameterf(m_textureTarget, GL_TEXTURE_MAX_ANISOTROPY_EXT, Clamp(0.0f, 8.0f, maxAnisotropy));
		}
		else
		{
			glTexParameteri(m_textureTarget, GL_TEXTURE_BASE_LEVEL, 0);
			glTexParameteri(m_textureTarget, GL_TEXTURE_MAX_LEVEL, 0);
		}
	}
}

void TextureData::InitRenderTargets(GLenum* attachments)
{
	if(attachments == 0)
		return;

	GLenum drawBuffers[32];      //32 is the max number of bound textures in OpenGL
	assert(m_numTextures <= 32); //Assert to be sure no buffer overrun should occur

	bool hasDepth = false;
	for(int i = 0; i < m_numTextures; i++)
	{
		if(attachments[i] == GL_DEPTH_ATTACHMENT)
		{
			drawBuffers[i] = GL_NONE;
			hasDepth = true;
		}
		else
			drawBuffers[i] = attachments[i];
	
		if(attachments[i] == GL_NONE)
			continue;
		
		if(m_frameBuffer == 0)
		{
			glGenFramebuffers(1, &m_frameBuffer);
			glBindFramebuffer(GL_FRAMEBUFFER, m_frameBuffer);
		}
		
		glFramebufferTexture2D(GL_FRAMEBUFFER, attachments[i], m_textureTarget, m_textureID[i], 0);
	}
	
	if(m_frameBuffer == 0)
		return;
	
	if(!hasDepth)
	{
		glGenRenderbuffers(1, &m_renderBuffer);
		glBindRenderbuffer(GL_RENDERBUFFER, m_renderBuffer);
		glRenderbufferStorage(GL_RENDERBUFFER, GL_DEPTH_COMPONENT, m_width, m_height);
		glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_RENDERBUFFER, m_renderBuffer);
	}
	
	glDrawBuffers(m_numTextures, drawBuffers);
	
	//glDrawBuffer(GL_NONE);
	//glReadBuffer(GL_NONE);
	
	if(glCheckFramebufferStatus(GL_FRAMEBUFFER) != GL_FRAMEBUFFER_COMPLETE)
	{
		std::cerr << "Framebuffer creation failed!" << std::endl;
		assert(false);
	}
	
	glBindFramebuffer(GL_FRAMEBUFFER, 0);
}

void TextureData::Bind(int textureNum) const
{
	glBindTexture(m_textureTarget, m_textureID[textureNum]);
}

void TextureData::BindAsRenderTarget() const
{
	glBindTexture(GL_TEXTURE_2D,0);
	glBindFramebuffer(GL_FRAMEBUFFER, m_frameBuffer);
	
	#if PROFILING_SET_1x1_VIEWPORT == 0
		glViewport(0, 0, m_width, m_height);
	#else
		glViewport(0, 0, 1, 1);
	#endif
}

Texture::Texture(const std::string& fileName, GLenum textureTarget, GLfloat filter, GLenum internalFormat, GLenum format, bool clamp, GLenum attachment)
{
    std::string filePath = fileName;
    if (filePath.find(":") == std::string::npos)
        filePath = Util::ResourcePath() + "textures/" + fileName;
 	m_fileName = filePath;

	std::map<std::string, TextureData*>::const_iterator it = s_resourceMap.find(filePath);
	if(it != s_resourceMap.end())
	{
		m_textureData = it->second;
		m_textureData->AddReference();
	}
	else
	{
		int x, y, bytesPerPixel;
        
		unsigned char* data = stbi_load((filePath).c_str(), &x, &y, &bytesPerPixel, 4);

		if(data == NULL)
		{
			std::cerr << "Unable to load texture: " << filePath << std::endl;
		}

		m_textureData = new TextureData(textureTarget, x, y, 1, &data, &filter, &internalFormat, &format, clamp, &attachment);
		stbi_image_free(data);
		
		s_resourceMap.insert(std::pair<std::string, TextureData*>(filePath, m_textureData));
	}
}

Texture::Texture(int width, int height, unsigned char* data, GLenum textureTarget, GLfloat filter, GLenum internalFormat, GLenum format, bool clamp, GLenum attachment)
{
	m_fileName = "";
	m_textureData = new TextureData(textureTarget, width, height, 1, &data, &filter, &internalFormat, &format, clamp, &attachment);
}

Texture::Texture(const Texture& texture) :
	m_textureData(texture.m_textureData),
	m_fileName(texture.m_fileName)
{
	m_textureData->AddReference();
}

void Texture::operator=(Texture texture)
{
	char* temp[sizeof(Texture)/sizeof(char)];
	memcpy(temp, this, sizeof(Texture));
	memcpy(this, &texture, sizeof(Texture));
	memcpy(&texture, temp, sizeof(Texture));
}

Texture::~Texture()
{
	if(m_textureData && m_textureData->RemoveReference())
	{
		if(m_fileName.length() > 0)
			s_resourceMap.erase(m_fileName);
			
		delete m_textureData;
	}
}

void Texture::Bind(unsigned int unit) const
{
	assert(unit >= 0 && unit <= 31);
	glActiveTexture(GL_TEXTURE0 + unit);
	m_textureData->Bind(0);
}

void Texture::BindAsRenderTarget() const
{
	m_textureData->BindAsRenderTarget();
}
